### PR TITLE
Design doc for RestoreItemAction wait for AdditionalItems to be ready

### PR DESCRIPTION
**Pull request by sseago**
Design for https://github.com/vmware-tanzu/velero/issues/1350

### CODE REVIEW

File: design/wait-for-additional-items.md
Comment (alaypatel07) [inline design/wait-for-additional-items.md:22]: s/additioonal/additional

**Comment by sseago** [comment]
I'm going to update this shortly with a third implementation option which should get rid of the need for making interface changes. Instead of adding a new func to either an existing or new interface, we could replace the `WaitForAdditionalItems` bool in my proposed changes to `RestoreItemActionExecuteOutput` with `AdditionalItemsReadyFunc` which would be a pointer to a func. If a plugin ignores it (as would any pre-existing plugins), it remains nil and velero doesn't attempt to wait. If a plugin needs to wait for AdditionalItems to be ready, instead of setting an entry in the return struct to 'true', it will return a func pointer, which will be the func that velero's `restoreItem` func will call in its wait method rather than a named interface func. This will remove the need for existing plugins to have to make changes here without introducing optional interfaces.

**Comment by nrb** [comment]
Thanks @sseago! I know you're working on this, so I'm likely to review after the v1.5.0 release and you're a little further along. I'll also be on vacation from Sept 4 to Sept 14th, so it's probably going to be after that, just to set expectations.

One question I do have - are you/your team able to sign up for implementation of this once approved?

**Comment by sseago** [comment]
@nrb Yes, my plan was to put together a PR to implement this once we got sign-off on the design. I expect one more update to the design before it's ready to review -- to propose a third alternative which I think will eliminate both the backwards-compatibilty problem and the need to define a new optional interface, as I mention in the above comment. I may get to it today, or tomorrow at the latest. Either way I assume your current priority is getting 1.5.0 out the door.

**Review by nrb — COMMENTED** [review]
So far I think this is reasonable.

I like the idea of making the `AdditionalItemsReadyFunc` a pointer so that not all RestoreItemAction plugins have to implement it.

I do still have a few questions on specific areas, and would like to hear from the rest of @vmware-tanzu/velero-maintainers

File: design/wait-for-additional-items.md
Comment (sseago) [reply design/wait-for-additional-items.md:53]: Yes. I was thinking of something very similar to `crdAvailable`, defined in the same file. Something like this (logging, error checking removed to focus on func operation):
```
// crdAvailable waits for additionalItems to be available for use before letting the restore continue.
func (ctx *restoreContext) itemsAvailable(func readyFunc, items []ResourceIdentifier, crdClient client.Dynamic) (bool, error) {
	var available bool
	err := wait.PollImmediate(time.Second, ctx.resourceTerminatingTimeout, func() (bool, error) {
		available, err = readyFunc(ctx.restore, items)
		if err != nil {
			return true, err
		}
		return available, nil
	})
	return available, err
}
```

File: design/wait-for-additional-items.md
Comment (sseago) [reply design/wait-for-additional-items.md:99]: I didn't include it here since it's not part of the struct itself, but I can add it to the doc. It would look like this:
```
// WithoutRestore returns SkipRestore for RestoreItemActionExecuteOutput
func (r *RestoreItemActionExecuteOutput) WithoutItemsWait(readyFunc func( *api.Restore, []ResourceIdentifier)) *RestoreItemActionExecuteOutput {
        r.AdditionalItemsReadyFunc = readyFunc
        return r
}
```

File: design/wait-for-additional-items.md
Comment (sseago) [reply design/wait-for-additional-items.md:80]: `WithItemsWait` will be defined in restore_item_action.go. To use the example from the design doc, here's how that function would be used (with a custom plugin-defined `AreItemsReady` function. The following code would be in the custom plugin code:
```
func AreItemsReady (restore *api.Restore, additionalItems []ResourceIdentifier) (bool, error) {
	...
	return true, nil
}
func (p *RestorePlugin) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
	...
	return velero.NewRestoreItemActionExecuteOutput(input.Item).WithItemsWait(AreItemsReady), nil
}
```

Comment (nrb) [reply design/wait-for-additional-items.md:80]: Thanks! I was looking for where the definition would be versus the example usage, but you provided it in a reply to another comment.

File: design/wait-for-additional-items.md
Comment (nrb) [reply design/wait-for-additional-items.md:99]: I think it would be worth including as it's part of the change - I was wondering where the actually implementation would live. It wasn't clear to me from the text, hence having 2 questions about it :)

**Comment by sseago** [comment]
Updated with the `WithItemsWait` func definition added to the document.

**Comment by nrb** [comment]
@vmware-tanzu/velero-maintainers Please give this a review.

File: design/wait-for-additional-items.md
Comment (ashish-amarnath) [inline design/wait-for-additional-items.md:75]: Would it be simpler to have a restore level option of `--sync-restore-additional-items`, a new field in the `RestoreSpec`, which if set, we would call a `waitForAdditionalItems` function after finishing the `for` loop at https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L1026-L1056
The `waitForAdditionalItems` would use the `wait.PollImmediate` API to wait for the additional items to be ready. The time out for this can also be made configurable either using another restore level option, say, `--sync-restore-additional-items-timeout`.

The simplicity coming from:
1. With a new release of Velero, after this is implemented, all RestoreItemActions will be able to take advantage of this feature without having to make any changes to their plugins.
1. Current `RestoreItemAction` would not require to be updated. Similar to the limitation of an earlier iteration of the design.
1. The proposed design here, would require a release of Velero, with the updated `RestoreItemActionOutput` API type before plugins are able to release a version using the new fields in the `RestoreItemActionOutput`
1. Implementation of the `AdditionalItemsReadyFunc` would be very identical across all RestoreItemAction plugins. So, instead of having the same implementation across many plugins, this function may be implemented as the `waitForAdditionalItems` and unexported function in the restore package of core Velero.

File: design/wait-for-additional-items.md
Comment (ashish-amarnath) [inline design/wait-for-additional-items.md:59]: `resourceTerminatingTimeout` is not meant to be used for this purpose. Suggest creating a different constant/config for this.

**Review by ashish-amarnath — COMMENTED** [review]
@sseago Thanks for this proposal and apologies for not looking at this sooner.
I added a couple of comments. PTAL. Happy to discuss further either in the PR or in a separate design meeting. I can schedule a design meeting if that is preferred.

File: design/wait-for-additional-items.md
Comment (alaypatel07) [reply design/wait-for-additional-items.md:75]: @ashish-amarnath interesting thought but I have a question:

In the solution you proposed, how would `waitForAdditionalItems` be implemented in a generic way for all resources?

The concept of "Readiness" may be different for different resources (for example for a pod it may be the container ready status, for a secret, it might be the meager existence and for a custom resource it might be a custom field in the status), and we need the plugin for a particular resource to defined what Readiness means for the resource it needs to wait.

Comment (ashish-amarnath) [reply design/wait-for-additional-items.md:75]: @alaypatel07 Yes, `waitForAdditionalItems` is generic across all resources. This function would wait for the restored resource to be present in the response to the `Get` from the API server.
I interpreted "ready" as "existence". It appeared that was the race that issue #1350 was referring to, that, restored resources were not immediately available.

Comment (sseago) [reply design/wait-for-additional-items.md:75]: @ashish-amarnath So the basic reason for the plugin-specific function was that in discussing it with @nrb we weren't sure whether it was appropriate to attempt "is it ready?" quite as simply as "the Get API returns true". If we combined your proposal with keeping `AdditionalItemsReadyFunc` as an optional field where the generic solution fails, though, that might be workable.

Comment (sseago) [reply design/wait-for-additional-items.md:75]: Another possible concern was that since most resource types don't actually need this waiting -- in most cases, the Kubernetes model of eventual consistency means that it's OK to restore the current item even if the `additionalItems` aren't available yet, so this would be needless waiting. Having the plugin signal when to wait makes this the exception rather than the rule. Since we weren't using a generic Get call for "is it ready?" we were able to combine "this plugin waits for additional items" with "this is how we wait" in one field. If we wanted to have the generic function be the default with `AdditionalItemsReadyFunc` as an override for how we wait, we could implement optional waiting with an additional boolean field here. That way we'd only wait when we're dealing with a resource which will fail restore if it's not yet there.  Then, once we have this (either as-is in the proposal, or with this new boolean field to allow for plugins not to have to redefine the method when the generic `Get` approach is good enough), we may not need the `--sync-restore-additional-items` option, although we could provide it -- if it's true, we wait for items when the plugin indicates we should, and if its' false, we never wait. But at this point, I don't know whether that option provides value. The point of having this in the first place is that with certain plugins/resources, not waiting will fail migrations, and with others waiting doesn't help. It feels like a plugin-specific setting rather than a restore-specific settings.

File: design/wait-for-additional-items.md
Comment (sseago) [reply design/wait-for-additional-items.md:59]: I agree that a different method of determining it would be better. I didn't include that in the design as I wasn't sure what made more sense. We can include it as a restore option as suggested above, although we'll need to provide a sensible default (perhaps via a new const). We could also allow plugin-level overriding, but it's not obvious whether or not we need that additional complexity and if we add it, what the priority would be for using. Perhaps first priority is to use the value set on the restore CR, if provided. Then plugin-level value (in the plugin `RestoreItemExecuteOutput` struct), and if both are empty, use something defined in a const.

File: design/wait-for-additional-items.md
Comment (nrb) [reply design/wait-for-additional-items.md:75]: @ashish-amarnath Just waiting for `GET` isn't quite enough for readiness all the time. Readiness means the thing exists and is useful.

As an example, CRDs can exist before CRDs can be made. Readiness is determined by examining `status.conditions`. [See this logic](https://github.com/vmware-tanzu/velero/blob/dbc83af77b0c0a36a27ac1646b6504fa78bedaba/pkg/util/kube/utils.go#L202-L209)

As another example, a `Pod` can exist but be in `CrashLoopBackOff`, and therefore not ready.

File: design/wait-for-additional-items.md
Comment (zubron) [reply design/wait-for-additional-items.md:59]: I think I'm missing something but what would it look like for the timeout value to be configurable at the plugin level? Is it possible under the current plugin model for specific configuration to be passed in when the plugins are being added? Or did you mean that different plugins would define how long they would wait but it would be a fixed duration?

Comment (sseago) [reply design/wait-for-additional-items.md:59]: So one way to do it which avoids having to change any existing plugins that don't need the new functionality is to add an additional (optional) timeout field to the `RestoreItemActionExecuteOutput` struct. When the `Execute` func returns, it could add this from a const defined within the plugin implementation. We already have one field we're adding to this struct to define the "wait for ready" func, and based on recent review comments, we may be adding another bool field which velero will use to decide whether to wait at all for this plugin (instead of overloading the func return value for this, to allow for a default "does `Get` return it?" implementation. That's what I was considering for the next update.

File: design/wait-for-additional-items.md
Comment (ashish-amarnath) [reply design/wait-for-additional-items.md:100]: This comment can be resolved?

File: design/wait-for-additional-items.md
Comment (ashish-amarnath) [reply design/wait-for-additional-items.md:75]: Resolving this comment.

File: design/wait-for-additional-items.md
Comment (ashish-amarnath) [reply design/wait-for-additional-items.md:81]: This can also be resolved.

**Review by ashish-amarnath — COMMENTED** [review]
This LGTM.
The only pending comment is about adding a timeout field in the `RestoreItemActionExecuteOutput` struct.

**Comment by sseago** [comment]
@ashish-amarnath I've updated the timeout section to include a new server-wide value specified in `serverConfig.additonalItemsReadyTimeout` (with a default const), as well as a plugin-level override value as an optional field in `RestoreItemActionExecuteOutput`.

**Review by zubron — APPROVED** [review]
Thanks for your time and effort on this, @sseago! I think this proposed design looks good and I don't have any additional questions or comments. Thanks for addressing them all! :+1:

**Review by ashish-amarnath — COMMENTED** [review]
This LGTM! Thanks for addressing the comments.
Leaving the merge to @nrb

**Review by carlisia — COMMENTED** [review]
This lgtm. Deferring to @nrb.

**Review by nrb — APPROVED** [review]
Thanks for your effort and patience on this @sseago!

**Pull request merged**
